app/src/main/java/org/p2p/wallet/user/repository/UserTokensDatabaseRepository.kt

```diff
@@ -0,0 +1,142 @@
+import org.p2p.wallet.home.model.TokenComparator
+import org.p2p.wallet.home.model.TokenConverter
+
+class UserTokensDatabaseRepository(
+    private val tokenServiceEventManager: TokenServiceEventManager,
+    private val userLocalRepository: UserLocalRepository,
+    private val tokensDao: TokenDao,
+    private val tokenConverter: TokenConverter,
+    private val coroutineDispatchers: CoroutineDispatchers,
+    private val tokenServiceEventPublisher: TokenServiceEventPublisher
+) : UserTokensLocalRepository, CoroutineScope {
+    override val coroutineContext: CoroutineContext = coroutineDispatchers.io
+
+    init {
+        tokenServiceEventManager.subscribe(TokenServiceEventSubscriber(::updatePricesForTokens))
+    }
+
+    override suspend fun updateTokens(tokens: List<Token.Active>) {
+        tokensDao.insertOrUpdate(tokens.map(tokenConverter::toDatabase))
+    }
+
+    override suspend fun updateUserToken(
+        newBalanceLamports: BigInteger,
+        mintAddress: Base58String,
+        publicKey: Base58String
+    ) {
+        val tokenToUpdate = findTokenByMintAddress(mintAddress)
+            ?.let { createUpdatedToken(it, newBalanceLamports) }
+            ?: createNewToken(mintAddress, newBalanceLamports, publicKey)
+
+        if (tokenToUpdate != null) {
+            tokensDao.insertOrUpdate(
+                entities = listOf(tokenConverter.toDatabase(tokenToUpdate))
+            )
+        }
+    }
+
+    override suspend fun removeIfExists(publicKey: String, symbol: String) {
+        tokensDao.removeIfExists(publicKey, symbol)
+    }
+
+    override suspend fun findTokenByMintAddress(mintAddress: Base58String): Token.Active? {
+        return tokensDao.findByMintAddress(mintAddress.base58Value)
+            ?.let(tokenConverter::fromDatabase)
+    }
+
+    override fun observeUserTokens(): Flow<List<Token.Active>> {
+        return tokensDao.getTokensFlow()
+            .map { tokenEntities ->
+                tokenEntities.map(tokenConverter::fromDatabase)
+                    .sortedWith(TokenComparator())
+            }
+    }
+
+    override fun observeUserToken(mintAddress: Base58String): Flow<Token.Active> {
+        return observeUserTokens().distinctUntilChanged()
+            .map { tokens -> tokens.findByMintAddress(mintAddress.base58Value) }
+            .filterNotNull()
+    }
+
+    override suspend fun getUserTokens(): List<Token.Active> {
+        return tokensDao.getTokens()
+            .map(tokenConverter::fromDatabase)
+    }
+
+    override suspend fun clear() {
+        tokensDao.clearAll()
+    }
+
+    private fun updatePricesForTokens(prices: List<TokenServicePrice>) {
+        launch {
+            val oldTokens = getUserTokens()
+            val newTokens = oldTokens.map { token ->
+                val tokenRate = prices.firstOrNull { token.mintAddress == it.address }?.usdRate
+                token.copy(rate = tokenRate, totalInUsd = tokenRate?.let { token.total.times(it) })
```

Comment:
what if you just keep only needed data for user token like mint address and public key, and then collect data from prices and metadata when you fetch it? so there only one source of truth for prices and other stuff